### PR TITLE
Fix range of matched statements

### LINKED ISSUE #513: Range of matched statement patterns is wrong
https://sgrep.live/54X?version=experimental

Why `$VAL = '*' ` is not strictly respected and it matches everything else as if `$VAL` is just an un assigned metavar. The matching should stop at `res.set('access-control-allow-origin', x);`

### PR DESCRIPTION
**Pull request by aryx**

### CODE REVIEW

**Comment by aryx** [comment]
This probably also closes #513

**Comment by aryx** [comment]
This probably also closes #593

File: semgrep-core/matching/Generic_vs_generic.ml [hunk 1]
```diff
@@ -84,6 +84,18 @@ let m_string_xhp_text sa sb =
   else fail ()
 (*e: function [[Generic_vs_generic.m_string_xhp_text]] *)
 
+(* ugly, see comment in Semgrep_generic.match_sts_sts *)
+let env_add_matched_stmt st tin =
+  let key = "!STMTS!" in
```
Comment (nbrahms) [inline semgrep-core/matching/Generic_vs_generic.ml]: Can we make this an externally defined constant?

FWIW, I'd feel a bit more comfortable if we made the key type for the environment association a variant, but I realize that may be an annoyingly large change to make here.

File: semgrep-core/matching/Semgrep_generic.ml [hunk 2]
```diff
@@ -72,7 +72,35 @@ let match_st_st pattern e =
...
+   * TODO? do we need to generate unique key? we don't want
+   * nested calls to m_stmts_deep to polluate our metavar? We need
+   * to pass the key to m_stmts_deep?
+   *)
+  let key = "!STMTS!" in
+  let env = (key, Ss [])::env in
+
+  let res = GG.m_stmts_deep pattern e env in
+
+  res |> List.map (fun tin ->
```
Comment (nbrahms) [inline semgrep-core/matching/Semgrep_generic.ml:98]: Can you comment as to why we're reversing the statement list?

**Review by nbrahms — APPROVED** [review]
I'm very excited for this change; I think it will improve the quality of `semgrep` matches greatly.

I left a couple notes, but approving here for speed.

**Comment by ievans** [comment]
Would it be possible to address https://github.com/returntocorp/semgrep/issues/820 in this too @aryx ?

**Comment by aryx** [comment]
I agree with all your points @nbrahms Thx for the review! I like it! Leaving more comments.

**Pull request merged**